smarts/env/gymnasium/utils/action_conversion.py

```diff
@@ -0,0 +1,254 @@
+            space=_lane_space,
+            formatting_func=_format_lane_space,
+        ),
+        ActionSpaceType.LaneWithContinuousSpeed: _FormattingGroup(
+            space=_lane_with_continuous_speed_space,
+        ),
+        ActionSpaceType.MPC: _FormattingGroup(
+            space=_mpc_space,
+        ),
+        ActionSpaceType.MultiTargetPose: _FormattingGroup(
+            space=_multi_target_pose_space,
+        ),
+        ActionSpaceType.RelativeTargetPose: _FormattingGroup(
+            space=_relative_target_pose_space,
+        ),
+        ActionSpaceType.TargetPose: _FormattingGroup(
+            space=_target_pose_space,
+        ),
+        ActionSpaceType.Trajectory: _FormattingGroup(
+            space=_trajectory_space,
+        ),
+        ActionSpaceType.TrajectoryWithTime: _FormattingGroup(
+            space=_trajectory_with_time_space,
+        ),
+    }
+
+
+class ActionOptions(IntEnum):
+    """Defines the options for how the formatting matches the action space."""
+
+    multi_agent = 0
+    """Action must map to partial action space. Only active agents are included."""
+    full = 1
+    """Action must map to full action space. Inactive and active agents are included."""
+    unformatted = 2
+    """Actions are not reformatted or constrained to action space. Actions must directly map to
+    underlying SMARTS actions."""
+    default = 0
+    """Defaults to :attr:`multi_agent`."""
+
+
+class ActionSpacesFormatter:
+    """Formats actions to adapt SMARTS to `gym` environment requirements."""
+
+    def __init__(
+        self, agent_interfaces: Dict[str, AgentInterface], action_options: ActionOptions
+    ) -> None:
+        self._agent_interfaces = agent_interfaces
+        self._action_options = action_options
+
+        for agent_id, agent_interface in agent_interfaces.items():
+            assert self.supported(agent_interface.action), (
+                f"Agent `{agent_id}` is using an "
+                f"unsupported `{agent_interface.action}`."
+                f"Available actions:\n{json.dumps(set(agent_interfaces.keys()), indent=2)}"
+            )
+
+    def format(self, actions: Dict[str, Any]):
+        """Format the action to a format that SMARTS can use."""
+
+        if self._action_options == ActionOptions.unformatted:
+            return actions
+
+        out_actions = {}
+        formatting_groups = get_formats()
+        for agent_id, action in actions.items():
+            agent_interface = self._agent_interfaces[agent_id]
+            format_ = formatting_groups[agent_interface.action]
+            assert format_.space.contains(
+                action
+            ), f"Action {action} does not match space {format_.space}!"
+            formatted_action = format_.formatting_func(action)
+            out_actions[agent_id] = formatted_action
+
+        if self._action_options == ActionOptions.full:
+            assert self.space.contains(out_actions)
```

Comment:
+ If `self._action_options == ActionOptions.full`, all the active and inactive agent action spaces were checked in the preceeding `for` loop.
+ So instead of checking `assert self.space.contains(out_actions)` here, is a simple check whether `actions.keys()` matches `self._agent_interfaces.keys()` be sufficient here?

Reply:
From what I can see I did not actually check for inactive agents in the former loop. This makes sense because I can only check for actions that exist.

However, as you mention it makes more sense to just do a check on the `agent_ids` to test if the action is fully utilised instead of using `space.contains` because in every case where the `agent_ids` are fully utilised, they would have already been checked in the previous loop.

Reply:
I am comparing the keys instead.